### PR TITLE
Add to blacklist on resource page

### LINKED ISSUE #1706: UI for adding and removing users from the blacklist
Edit functionality for blacklist. The API exists already. This should be in the administration pages.

- [x] users can be added and removed to a per-resource blacklist by handlers

### PR DESCRIPTION
**Pull request by luontola**
Closes #1706

add

# Definition of Done / Review checklist

## Reviewability
- [x] link to issue
- [x] consider adding screenshots for ease of review

## API
- [x] API is documented and shows up in Swagger UI
- [x] API is backwards compatible or completely new
- [x] Events are backwards compatible

## Documentation
- [x] update changelog if necessary

## Testing
- [x] complex logic is unit tested
- [x] valuable features are integration / browser / acceptance tested automatically

## Accessibility
- [x] all fields have a label
- [ ] errors are linked to fields with aria-describedby
- [x] conscious decision about where to move focus after an action

## Follow-up
- [ ] new tasks are created for pending or remaining tasks
- [x] no critical TODOs left to implement

### CODE REVIEW

File: src/cljs/rems/administration/blacklist.cljs [hunk 1]
```diff
@@ -26,25 +27,100 @@
...
+
+(defn add-user-form [resource]
+  (let [user-field-id "blacklist-user"
+        all-users @(rf/subscribe [::all-users])
+        selected-users @(rf/subscribe [::selected-user])]
+    [:form.form-inline
+     {:on-submit (fn [event]
+                   (.preventDefault event)
+                   ;; TODO: more fancy input form for the comment?
+                   (when-some [comment (js/prompt (text :t.administration/comment))]
```
Comment (opqdonut) [inline src/cljs/rems/administration/blacklist.cljs]: yes please

File: src/cljs/rems/administration/blacklist.cljs [hunk 2]
```diff
@@ -26,25 +27,100 @@
...
+ ::selected-user
+ (fn [db _]
+   (::selected-user db)))
+
+(defn add-user-form [resource]
+  (let [user-field-id "blacklist-user"
+        all-users @(rf/subscribe [::all-users])
+        selected-users @(rf/subscribe [::selected-user])]
+    [:form.form-inline
+     {:on-submit (fn [event]
```
Comment (opqdonut) [inline src/cljs/rems/administration/blacklist.cljs]: we haven't used on-submit callbacks so far, but instead on-click callbacks on buttons. any reason for an on-submit here?

File: src/cljs/rems/administration/blacklist.cljs [hunk 3]
```diff
@@ -26,25 +27,100 @@
              :error-handler (flash-message/default-error-handler :top description)}))
    {:db (assoc db ::loading? true)}))
 
+(rf/reg-event-fx
+ ::add-to-blacklist
+ (fn [{:keys [db]} [_ resource user comment]]
+   (let [description [text :t.administration/add]]
+     (post! "/api/blacklist/add"
+            {:params {:blacklist/resource (select-keys resource [:resource/ext-id])
+                      :blacklist/user (select-keys user [:userid])
```
Comment (opqdonut) [inline src/cljs/rems/administration/blacklist.cljs]: good that `user` and `resource` are maps with standard keys ... but I wonder if there's a way to document this

File: src/cljs/rems/administration/blacklist.cljs (hunk 2 above)
Comment (luontola) [reply src/cljs/rems/administration/blacklist.cljs]: No big reason. That way it submits also by pressing enter in the user field.

File: src/clj/rems/db/blacklist.clj [hunk 4]
```diff
@@ -40,6 +42,30 @@
 (defn add-event! [event]
   (db/add-blacklist-event! {:eventdata (event->json event)}))
 
+(defn- command->event [command user-id]
```
Comment (opqdonut) [inline src/clj/rems/db/blacklist.clj]: when I wrote this stuff I thought rems.db.blacklist would only expose an event-level api, and the command could stay in the rems.api.blacklist ns

File: src/clj/rems/db/blacklist.clj [hunk 5]
```diff
@@ -40,6 +42,30 @@
...
+  {:event/actor user-id
+   :event/time (time/now)
+   :userid (get-in command [:blacklist/user :userid])
+   :resource/ext-id (get-in command [:blacklist/resource :resource/ext-id])
+   :event/comment (:comment command)})
+
+(defn- validate-event [event]
```
Comment (opqdonut) [inline src/clj/rems/db/blacklist.clj]: we have `validate-event` and on the other hand we have `(s/validate BlacklistEvent event)`... is there a risk of confusion between these two types of validation? should one be called something else?

File: src/clj/rems/db/blacklist.clj [hunk 6]
```diff
@@ -40,6 +42,30 @@
 (defn add-event! [event]
```
Comment (opqdonut) [inline src/clj/rems/db/blacklist.clj]: should this be private now? if it's private can it just be inlined?

File: src/clj/rems/db/blacklist.clj (hunk 4 above)
Comment (luontola) [reply src/clj/rems/db/blacklist.clj]: Ok, I can move them back to the other ns.

File: src/clj/rems/db/blacklist.clj [hunk 7]
```diff
@@ -49,22 +57,15 @@
...
-    (throw (IllegalArgumentException. "user doesn't exist")))
-  (when-not (resource/ext-id-exists? (:resource/ext-id event))
-    (throw (IllegalArgumentException. "resource doesn't exist")))
-  event)
-
 (defn add! [user-id command]
   (add-event! (-> (command->event command user-id)
                   (assoc :event/type :blacklist.event/add)
-                  validate-event)))
+                  check-foreign-keys)))
```
Comment (opqdonut) [inline src/clj/rems/db/blacklist.clj]: this is now redundant

**Pull request merged**